core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java

```diff
@@ -419,6 +410,16 @@ private void handleRequest(ChannelHandlerContext ctx, DecodedHttpRequest req) th
                     req.close();
                 }
             }
+
+            serviceResponse = serviceResponse.recover(cause -> {
```

Comment:
We didn't wrap the response returned by a service before. What'd be the performance impact of this? Could you share some benchmark results?

Reply:
I also considered the performance impact. 

> Could you share some benchmark results?

Sure, let me check the performance with some success and failure cases.

Reply:
I ran benchmarks and the working branch seems like having 2~3% overhead.
I am going to profile and find room for optimization.
* Working branch
  * HttpService
  ```
  go-wrk -M GET -c 10 -d 30 http://127.0.0.1:8080/http
  Running 30s test @ http://127.0.0.1:8080/http
    10 goroutine(s) running concurrently
  1194483 requests in 29.492190824s, 151.51MB read
  Requests/sec:		40501.67
  Transfer/sec:		5.14MB
  Avg Req Time:		246.903µs
  Fastest Request:	58.389µs
  Slowest Request:	128.902189ms
  Number of Errors:	0
  ```
  * Annotated service
	```
	go-wrk -M GET -c 10 -d 30 http://127.0.0.1:8080/anno/
	Running 30s test @ http://127.0.0.1:8080/anno/
	  10 goroutine(s) running concurrently
	1143563 requests in 29.512220953s, 145.05MB read
	Requests/sec:		38748.80
	Transfer/sec:		4.91MB
	Avg Req Time:		258.072µs
	Fastest Request:	60.938µs
	Slowest Request:	126.904941ms
	Number of Errors:	0
	```
* Master branch
	* HttpService
  ```
  go-wrk -M GET -c 10 -d 30 http://127.0.0.1:8080/http
  Running 30s test @ http://127.0.0.1:8080/http
    10 goroutine(s) running concurrently
  1220807 requests in 29.477914933s, 154.85MB read
  Requests/sec:		41414.29
  Transfer/sec:		5.25MB
  Avg Req Time:		241.462µs
  Fastest Request:	57.35µs
  Slowest Request:	125.217814ms
  Number of Errors:	0
  ```
	* Annotated service
  ```
  go-wrk -M GET -c 10 -d 30 http://127.0.0.1:8080/anno/
  Running 30s test @ http://127.0.0.1:8080/anno/
    10 goroutine(s) running concurrently
  1212290 requests in 29.479147698s, 153.77MB read
  Requests/sec:		41123.64
  Transfer/sec:		5.22MB
  Avg Req Time:		243.169µs
  Fastest Request:	57.252µs
  Slowest Request:	64.730558ms
  Number of Errors:	0
  ```

Reply:
When HttpHeaders cache(#3714) is applied, `HttpResponse.recover()` does not seem to have performance regression.
Some benchmarks show better performance than the master branch, thanks to  HttpHeaders cache(#3714).
| branch | service | Requests/sec | Transfer/sec | Avg Req Time | Fastest Request | Slowest Request | Number of Errors | 
| ------ | ------- | ------------ | ------------ | ------------ | --------------- | --------------- | ---------------- |  
| recover | HttpService | 42926.76 | 5.44MB | 232.954µs | 57.712µs | 108.570215ms | 0|
| recover | HttpService | 42353.22 | 5.37MB | 236.109µs | 54.949µs | 114.236967ms | 0|
| recover | HttpService | 43605.76 | 5.53MB | 229.327µs | 57.688µs | 124.909328ms | 0|
| recover | Annotated service | 41665.45 | 5.28MB | 240.007µs | 60.026µs | 84.9303ms|
| recover | Annotated service | 41119.65 | 5.22MB | 243.192µs | 57.256µs | 126.482998ms | 0|
| recover | Annotated service | 41406.57 | 5.25MB | 241.507µs | 59.819µs | 125.234891ms | 0|
| master | HttpService | 42112.26 | 5.34MB | 237.46µs | 54.834µs | 127.838011ms | 0|
| master | HttpService | 42256.39 | 5.36MB | 236.65µs | 56.247µs | 110.931958ms | 0|
| master | HttpService | 42108.98 | 5.34MB | 237.479µs | 58.551µs | 123.010129ms | 0|
| master | Annotated service | 41324.97 | 5.24MB | 241.984µs | 59.693µs | 114.15864ms | 0|
| master | Annotated service | 40893.37 | 5.19MB | 244.538µs | 60.023µs | 82.101197ms | 0|
| master | Annotated service | 41360.49 | 5.25MB | 241.776µs | 59.42µs | 103.278116ms | 0|